internal/pkg/service/common/etcdop/etcdop.go

```diff
@@ -16,101 +29,182 @@ import (
+	return BoolOp{opFactory: factory, processor: processor}
 }
 
 // NewGetOneOp wraps an operation, the result of which is one KV pair.
-func NewGetOneOp(etcdOp etcd.Op, processor getOneProcessor) GetOneOp {
-	return GetOneOp{op: op(etcdOp), processor: processor}
+func NewGetOneOp(factory opFactory, processor getOneProcessor) GetOneOp {
+	return GetOneOp{opFactory: factory, processor: processor}
+}
+
+// NewGetOneTOp wraps an operation, the result of which is one KV pair, value is encoded as the type T.
+func NewGetOneTOp[T any](factory opFactory, processor getOneTProcessor[T]) GetOneTOp[T] {
+	return GetOneTOp[T]{opFactory: factory, processor: processor}
 }
 
 // NewGetManyOp wraps an operation, the result of which is zero or multiple KV pairs.
-func NewGetManyOp(etcdOp etcd.Op, processor getManyProcessor) GetManyOp {
-	return GetManyOp{op: op(etcdOp), processor: processor}
+func NewGetManyOp(factory opFactory, processor getManyProcessor) GetManyOp {
+	return GetManyOp{opFactory: factory, processor: processor}
+}
+
+// NewGetManyTOp wraps an operation, the result of which is zero or multiple KV pairs, values are encoded as the type T.
+func NewGetManyTOp[T any](factory opFactory, processor getManyTProcessor[T]) GetManyTOp[T] {
+	return GetManyTOp[T]{opFactory: factory, processor: processor}
 }
 
 // NewCountOp wraps an operation, the result of which is a count.
-func NewCountOp(etcdOp etcd.Op, processor countProcessor) CountOp {
-	return CountOp{op: op(etcdOp), processor: processor}
+func NewCountOp(factory opFactory, processor countProcessor) CountOp {
+	return CountOp{opFactory: factory, processor: processor}
 }
 
 // NewNoResultOp wraps an operation, the result of which is an error or nil.
-func NewNoResultOp(etcdOp etcd.Op, processor noResultProcessor) NoResultOp {
-	return NoResultOp{op: op(etcdOp), processor: processor}
+func NewNoResultOp(factory opFactory, processor noResultProcessor) NoResultOp {
+	return NoResultOp{opFactory: factory, processor: processor}
+}
+
+// Op returns raw etcd.Op.
+func (v opFactory) Op(ctx context.Context) (*etcd.Op, error) {
+	return v(ctx)
 }
 
 func (v BoolOp) Do(ctx context.Context, client *etcd.Client) (result bool, err error) {
-	if r, err := client.Do(ctx, v.op.Op()); err == nil {
-		return v.processor(r)
+	if etcdOp, err := v.opFactory(ctx); err != nil {
+		return false, err
+	} else if r, err := client.Do(ctx, *etcdOp); err == nil {
+		return v.processor(ctx, r)
 	} else {
 		return false, err
 	}
 }
 
-func (v GetOneOp) Do(ctx context.Context, client *etcd.Client) (kv *mvccpb.KeyValue, err error) {
-	if r, err := client.Do(ctx, v.op.Op()); err == nil {
-		return v.processor(r)
+func (v GetOneOp) Do(ctx context.Context, client *etcd.Client) (kv *KeyValue, err error) {
+	if etcdOp, err := v.opFactory(ctx); err != nil {
+		return nil, err
+	} else if r, err := client.Do(ctx, *etcdOp); err == nil {
+		return v.processor(ctx, r)
+	} else {
+		return nil, err
+	}
+}
```

Comment:
Nasledujuce metody su rovnake, lisia sa iba v return values, to sa neda zapisat genericky.